File: auramind-gemini/android/app/src/main/java/com/auramind/app/AuraListenPlugin.java
package com.auramind.app;

import android.Manifest;
import android.content.Intent;
import android.os.Bundle;
import android.os.SystemClock;
import android.speech.RecognitionListener;
import android.speech.RecognizerIntent;
import android.speech.SpeechRecognizer;

import com.getcapacitor.JSObject;
import com.getcapacitor.PermissionState;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.annotation.PermissionCallback;

import java.util.ArrayList;

/**
 * AuraListen — native speech recognition for voice study.
 *
 * Android System WebView has no SpeechRecognition, so spoken answers never
 * worked in the Android app. This wraps android.speech.SpeechRecognizer and
 * reports the same shape of events the Web Speech API does, so
 * src/services/voice/nativeRecognition.ts can hand useVoiceStudy an object
 * that behaves like the browser's recogniser.
 *
 * Events carry the session id passed to start(), because listeners are
 * plugin-wide: a stop followed by a quick restart must not let the old
 * session's late "end" close the new one.
 *
 * SpeechRecognizer must be created and driven on the main thread.
 */
@CapacitorPlugin(
        name = "AuraListen",
        permissions = @Permission(strings = {Manifest.permission.RECORD_AUDIO}, alias = "microphone"))
public class AuraListenPlugin extends Plugin {

    private SpeechRecognizer recognizer;
    private String session = "";
    private long lastLevelAt = 0;

    @PluginMethod
    public void isAvailable(PluginCall call) {
        JSObject out = new JSObject();
        out.put("available", SpeechRecognizer.isRecognitionAvailable(getContext()));
        call.resolve(out);
    }

    @PluginMethod
    public void start(PluginCall call) {
        if (getPermissionState("microphone") != PermissionState.GRANTED) {
            requestPermissionForAlias("microphone", call, "microphonePermissionResult");
            return;
        }
        begin(call);
    }

    @PermissionCallback
    private void microphonePermissionResult(PluginCall call) {
        if (getPermissionState("microphone") == PermissionState.GRANTED) {
            begin(call);
        } else {
            call.reject("Microphone permission denied", "not-allowed");
        }
    }

    private void begin(PluginCall call) {
        if (!SpeechRecognizer.isRecognitionAvailable(getContext())) {
            call.reject("No speech recognition service on this device", "unavailable");
            return;
        }
        String nextSession = call.getString("session", "");
        String lang = call.getString("lang", "en-US");
        boolean partial = Boolean.TRUE.equals(call.getBoolean("interimResults", true));

        getActivity().runOnUiThread(() -> {
            // A new session replaces the old one, which is told it ended.
            endSession(true);
            session = nextSession;
            recognizer = SpeechRecognizer.createSpeechRecognizer(getContext());
            recognizer.setRecognitionListener(listener);

            Intent intent = new Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH);
            intent.putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, RecognizerIntent.LANGUAGE_MODEL_FREE_FORM);
            intent.putExtra(RecognizerIntent.EXTRA_LANGUAGE, lang);
            intent.putExtra(RecognizerIntent.EXTRA_PARTIAL_RESULTS, partial);
            intent.putExtra(RecognizerIntent.EXTRA_MAX_RESULTS, 1);
            intent.putExtra(RecognizerIntent.EXTRA_CALLING_PACKAGE, getContext().getPackageName());
            try {
                recognizer.startListening(intent);
                call.resolve();
            } catch (RuntimeException e) {
                endSession(false);
                call.reject("Could not start listening", "unknown");
            }
        });
    }

    /** Stop and deliver what was heard (like SpeechRecognition.stop()). */
    @PluginMethod
    public void stop(PluginCall call) {
        getActivity().runOnUiThread(() -> {
            if (recognizer != null) recognizer.stopListening();
            call.resolve();
        });
    }

    /** Stop and discard (like SpeechRecognition.abort()). */
    @PluginMethod
    public void abort(PluginCall call) {
        getActivity().runOnUiThread(() -> {
            endSession(true);
            call.resolve();
        });
    }

    /** Leaving the app releases the microphone rather than listening in the background. */
    @Override
    protected void handleOnPause() {
        getActivity().runOnUiThread(() -> endSession(true));
    }

    @Override
    protected void handleOnDestroy() {
        endSession(false);
    }

    /** Tears down the recogniser; announces "aborted" + "end" when asked. */
    private void endSession(boolean announce) {
        if (recognizer == null) return;
        SpeechRecognizer old = recognizer;
        recognizer = null;
        try {
            old.cancel();
            old.destroy();
        } catch (RuntimeException ignored) {
            // Service already gone.
        }
        if (announce) {
            emit("error", "error", "aborted");
            emit("end", null, null);
        }
    }

    private void emit(String event, String key, String value) {
        JSObject data = new JSObject();
        data.put("session", session);
        if (key != null) data.put(key, value);
        notifyListeners(event, data);
    }

    private static String firstResult(Bundle bundle) {
        if (bundle == null) return "";
        ArrayList<String> results = bundle.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION);
        return results == null || results.isEmpty() || results.get(0) == null ? "" : results.get(0);
    }

    /** SpeechRecognizer error codes, in Web Speech API vocabulary. */
    static String webError(int code) {
        switch (code) {
            case SpeechRecognizer.ERROR_NO_MATCH:
            case SpeechRecognizer.ERROR_SPEECH_TIMEOUT:
                return "no-speech";
            case SpeechRecognizer.ERROR_INSUFFICIENT_PERMISSIONS:
                return "not-allowed";
            case SpeechRecognizer.ERROR_AUDIO:
                return "audio-capture";
            case SpeechRecognizer.ERROR_NETWORK:
            case SpeechRecognizer.ERROR_NETWORK_TIMEOUT:
            case SpeechRecognizer.ERROR_SERVER:
            case 10: // ERROR_TOO_MANY_REQUESTS (API 31)
            case 11: // ERROR_SERVER_DISCONNECTED (API 31)
                return "network";
            case 12: // ERROR_LANGUAGE_NOT_SUPPORTED (API 31)
            case 13: // ERROR_LANGUAGE_UNAVAILABLE (API 31)
                return "language-not-supported";
            case SpeechRecognizer.ERROR_CLIENT:
                return "aborted";
            default:
                return "unknown";
        }
    }

    private final RecognitionListener listener = new RecognitionListener() {
        @Override
        public void onReadyForSpeech(Bundle params) {
            emit("start", null, null);
        }

        @Override
        public void onBeginningOfSpeech() { }

        @Override
        public void onRmsChanged(float rmsdB) {
            // ~10 updates a second is plenty for the orb and keeps the bridge quiet.
            long now = SystemClock.uptimeMillis();
            if (now - lastLevelAt < 100) return;
            lastLevelAt = now;
            // rmsdB runs roughly -2 (silence) to 10 (loud speech).
            float level = Math.max(0f, Math.min(1f, (rmsdB + 2f) / 12f));
            JSObject data = new JSObject();
            data.put("session", session);
            data.put("level", level);
            notifyListeners("level", data);
        }

        @Override
        public void onBufferReceived(byte[] buffer) { }

        @Override
        public void onEndOfSpeech() { }

        @Override
        public void onError(int error) {
            emit("error", "error", webError(error));
            emit("end", null, null);
            endSession(false);
        }

        @Override
        public void onResults(Bundle results) {
            emit("final", "text", firstResult(results));
            emit("end", null, null);
            endSession(false);
        }

        @Override
        public void onPartialResults(Bundle partialResults) {
            String text = firstResult(partialResults);
            if (!text.isEmpty()) emit("partial", "text", text);
        }

        @Override
        public void onEvent(int eventType, Bundle params) { }
    };
}

File: auramind-gemini/ios/App/App/AuraListenPlugin.swift
import AVFoundation
import Capacitor
import Foundation
import Speech

/// AuraListen — native speech recognition for voice study on iOS.
///
/// WKWebView exposes `webkitSpeechRecognition` but it never delivers results
/// (WebKit bug 239816), so spoken answers would silently fail in the iOS app.
/// This wraps SFSpeechRecognizer and reports the same events as the Android
/// plugin (android/.../AuraListenPlugin.java): start, partial, final, level,
/// error and end, each tagged with the session passed to start(), in Web
/// Speech API vocabulary. src/services/voice/nativeRecognition.ts turns them
/// into a browser-shaped recogniser for useVoiceStudy.
///
/// Android's recogniser ends by itself after a pause; SFSpeechRecognizer keeps
/// listening until told to stop, so a silence timer ends the utterance to
/// behave the same way.
@objc(AuraListenPlugin)
public class AuraListenPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "AuraListenPlugin"
    public let jsName = "AuraListen"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "isAvailable", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "start", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stop", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "abort", returnType: CAPPluginReturnPromise)
    ]

    /// Pause after the last recognised words that ends the utterance.
    private let endOfSpeechSilence: TimeInterval = 1.6
    /// How long to wait for any speech at all before reporting no-speech.
    private let noSpeechTimeout: TimeInterval = 8.0

    private let audioEngine = AVAudioEngine()
    private var recognizer: SFSpeechRecognizer?
    private var request: SFSpeechAudioBufferRecognitionRequest?
    private var task: SFSpeechRecognitionTask?
    private var session = ""
    private var heardText = ""
    private var finalSent = false
    private var active = false
    private var silenceTimer: Timer?
    private var lastLevelAt: TimeInterval = 0

    override public func load() {
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(appWillResignActive),
            name: UIApplication.willResignActiveNotification,
            object: nil
        )
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }

    /// Leaving the app releases the microphone rather than listening in the background.
    @objc private func appWillResignActive() {
        DispatchQueue.main.async { self.endSession(announce: true) }
    }

    @objc func isAvailable(_ call: CAPPluginCall) {
        let available = SFSpeechRecognizer(locale: Locale(identifier: "en-US"))?.isAvailable ?? false
        call.resolve(["available": available])
    }

    @objc func start(_ call: CAPPluginCall) {
        let nextSession = call.getString("session") ?? ""
        let lang = call.getString("lang") ?? "en-US"
        let partial = call.getBool("interimResults") ?? true

        requestPermissions { granted in
            guard granted else {
                call.reject("Microphone or speech recognition permission denied", "not-allowed")
                return
            }
            DispatchQueue.main.async {
                self.begin(call, session: nextSession, lang: lang, partial: partial)
            }
        }
    }

    /// Stop and deliver what was heard (like SpeechRecognition.stop()).
    @objc func stop(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            self.finishListening()
            call.resolve()
        }
    }

    /// Stop and discard (like SpeechRecognition.abort()).
    @objc func abort(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            self.endSession(announce: true)
            call.resolve()
        }
    }

    // MARK: - Permissions

    private func requestPermissions(_ done: @escaping (Bool) -> Void) {
        SFSpeechRecognizer.requestAuthorization { status in
            guard status == .authorized else {
                done(false)
                return
            }
            AVAudioSession.sharedInstance().requestRecordPermission { granted in
                done(granted)
            }
        }
    }

    // MARK: - Session

    private func begin(_ call: CAPPluginCall, session nextSession: String, lang: String, partial: Bool) {
        // A new session replaces the old one, which is told it ended.
        endSession(announce: true)

        guard let recognizer = SFSpeechRecognizer(locale: Locale(identifier: lang)) else {
            call.reject("Speech recognition does not support \(lang)", "language-not-supported")
            return
        }
        guard recognizer.isAvailable else {
            call.reject("Speech recognition is not available right now", "unavailable")
            return
        }

        session = nextSession
        heardText = ""
        finalSent = false
        self.recognizer = recognizer

        do {
            let audio = AVAudioSession.sharedInstance()
            try audio.setCategory(.playAndRecord, mode: .measurement, options: [.duckOthers, .defaultToSpeaker, .allowBluetooth])
            try audio.setActive(true, options: .notifyOthersOnDeactivation)
        } catch {
            call.reject("Could not open the microphone", "audio-capture")
            return
        }

        let request = SFSpeechAudioBufferRecognitionRequest()
        request.shouldReportPartialResults = partial
        self.request = request

        let input = audioEngine.inputNode
        let format = input.outputFormat(forBus: 0)
        guard format.sampleRate > 0, format.channelCount > 0 else {
            restoreAudioSession()
            call.reject("No microphone input is available", "audio-capture")
            return
        }
        input.removeTap(onBus: 0)
        input.installTap(onBus: 0, bufferSize: 1024, format: format) { [weak self] buffer, _ in
            self?.request?.append(buffer)
            self?.reportLevel(buffer)
        }

        audioEngine.prepare()
        do {
            try audioEngine.start()
        } catch {
            input.removeTap(onBus: 0)
            restoreAudioSession()
            call.reject("Could not start listening", "audio-capture")
            return
        }

        active = true
        let mine = nextSession
        task = recognizer.recognitionTask(with: request) { [weak self] result, error in
            DispatchQueue.main.async {
                guard let self = self, self.active, self.session == mine else { return }
                if let result = result {
                    let text = result.bestTranscription.formattedString
                    if !text.isEmpty { self.heardText = text }
                    if result.isFinal {
                        self.deliverFinal()
                        self.endSession(announce: false)
                        return
                    }
                    if !text.isEmpty {
                        self.emit("partial", ["text": text])
                        self.armSilenceTimer(self.endOfSpeechSilence)
                    }
                }
                if error != nil {
                    if self.heardText.isEmpty {
                        self.emit("error", ["error": "no-speech"])
                    } else {
                        self.deliverFinal()
                    }
                    self.endSession(announce: false)
                }
            }
        }

        emit("start", [:])
        armSilenceTimer(noSpeechTimeout)
        call.resolve()
    }

    /// Ends the audio and lets the recogniser deliver its final result.
    private func finishListening() {
        guard active else { return }
        silenceTimer?.invalidate()
        silenceTimer = nil
        if audioEngine.isRunning {
            audioEngine.stop()
            audioEngine.inputNode.removeTap(onBus: 0)
        }
        request?.endAudio()
        // If the recogniser never answers, close out with what was heard.
        let mine = session
        DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) { [weak self] in
            guard let self = self, self.active, self.session == mine else { return }
            if self.heardText.isEmpty {
                self.emit("error", ["error": "no-speech"])
            } else {
                self.deliverFinal()
            }
            self.endSession(announce: false)
        }
    }

    private func armSilenceTimer(_ interval: TimeInterval) {
        silenceTimer?.invalidate()
        silenceTimer = Timer.scheduledTimer(withTimeInterval: interval, repeats: false) { [weak self] _ in
            self?.finishListening()
        }
    }

    private func deliverFinal() {
        guard !finalSent else { return }
        finalSent = true
        if !heardText.isEmpty {
            emit("final", ["text": heardText])
        }
    }

    /// Tears everything down; announces "aborted" first when asked, and
    /// always announces "end" for a session that was running.
    private func endSession(announce: Bool) {
        guard active else { return }
        active = false
        silenceTimer?.invalidate()
        silenceTimer = nil
        if audioEngine.isRunning {
            audioEngine.stop()
        }
        audioEngine.inputNode.removeTap(onBus: 0)
        task?.cancel()
        task = nil
        request = nil
        recognizer = nil
        restoreAudioSession()
        if announce {
            emit("error", ["error": "aborted"])
        }
        emit("end", [:])
    }

    /// Back to playback so read-aloud is audible again after listening.
    private func restoreAudioSession() {
        let audio = AVAudioSession.sharedInstance()
        try? audio.setActive(false, options: .notifyOthersOnDeactivation)
        try? audio.setCategory(.playback, mode: .spokenAudio, options: [.duckOthers])
    }

    // MARK: - Events

    private func reportLevel(_ buffer: AVAudioPCMBuffer) {
        guard let samples = buffer.floatChannelData?[0] else { return }
        let count = Int(buffer.frameLength)
        guard count > 0 else { return }
        var sum: Float = 0
        for i in 0..<count {
            sum += samples[i] * samples[i]
        }
        let rms = sqrt(sum / Float(count))
        let db = 20 * log10(max(rms, 0.000_001))
        // -50 dB (quiet room) … -10 dB (speaking close to the mic) → 0 … 1
        let level = min(1, max(0, (db + 50) / 40))

        let now = Date().timeIntervalSince1970
        guard now - lastLevelAt >= 0.1 else { return }
        lastLevelAt = now
        DispatchQueue.main.async { [weak self] in
            guard let self = self, self.active else { return }
            self.emit("level", ["level": Double(level)])
        }
    }

    private func emit(_ event: String, _ data: [String: Any]) {
        var payload = data
        payload["session"] = session
        notifyListeners(event, data: payload)
    }
}
